cpp/ql/lib/semmle/code/cpp/commons/unix/Constants.qll

```diff
@@ -11,10 +11,10 @@ import cpp
  */
 bindingset[input]
 int parseOctal(string input) {
-  input.charAt(0) = "0" and
+  input.regexpMatch("0[0-7]+") and
   result =
-    strictsum(int ix |
-      ix in [0 .. input.length()]
+    sum(int ix |
+      ix in [1 .. input.length()]
```

Comment:
If the input matches the regexp `0[0-7]+` the length of the string must be at least 2, and so the range `[1 .. input.length()]` must be non-empty, right? So can't this still be a strictsum?

Reply:
It can be, but it's not necessary. Is there a reason to prefer `strictsum`?

Reply:
`strictsum` is more efficient 😍. `sum` needs to handle the case where the `[1 .. input.length()]` range is empty, while `strictsum` doesn't. I've included the gory details if you're curious:

This query:
```ql
from int n
where n = sum(Expr e | exists(e.getValue().toInt()) | e.getValue().toInt())
select n
```
compiles to the following DIL code:
```
noinline project##select#query#f#nullary() :-
  exists(cached string call_result#2 |
    exists(/* Expr::Expr */ cached dontcare entity _ |
      Expr::Expr::getValue_dispred#ff(_, call_result#2)
    ),
    exists(dontcare int _ | toInt(call_result#2, _))
  )
.

#select#query#f(unique int n) :-
  n = sum(int sum#project#)(/* Expr::Expr */ cached entity e, int sum#expr# |
    exists(cached string call_result#4 |
      Expr::Expr::getValue_dispred#ff(e, call_result#4),
      toInt(call_result#4, sum#expr#)
    ) |
    sum#project# = sum#expr#
  );
  (not(project##select#query#f#nullary()), n = 0)
.
```
The `(not(project##select#query#f#nullary()), n = 0)` disjunct is the special case that needs to be there in case the range is empty (if you look closely you'll see that `project##select#query#f#nullary()` is a predicate that holds if the range is non-empty).

Compare that to the DIL for this query:
```ql
from int n
where n = strictsum(Expr e | exists(e.getValue().toInt()) | e.getValue().toInt())
select n
```
here you get:
```
#select#query#f(unique int n) :-
  n = sum(int strictsum#project#)(/* Expr::Expr */ cached entity e,
                                  int strictsum#expr# |
    exists(/* @value */ cached entity v |
      exists(cached string call_result#4 |
        values(v, call_result#4), toInt(call_result#4, strictsum#expr#)
      ),
      valuebind(v, e)
    ) |
    strictsum#project# = strictsum#expr#
  )
.
```
which is basically the same piece of code, but with no checking if the range is non-empty.